sycl/doc/extensions/SubGroupBitmaskGroupFunctions/sub_group_bitmask_group_algorithms.asciidoc

```diff
@@ -0,0 +1,253 @@
+synchronized forward progress of work-items in the same sub-group. This
+divergence is why work-items following the same control flow may not be
+active at the same time. Care should be taken when modifying the value of the 
+returned `sub_group_mask` since this will change the behavior of masked group functions.
+
+`ext_oneapi_rank_in_mask` returns an integer id representing the calling work-item's
+position within sub-group subset represented by the provided `sub_group_mask`. This is
+undefined behavior if bit `n` of the `sub_group_mask` is not set, where `n` is the
+integer id of the work-item in the sub-group. As an example, if the bitmask was 
+`0b00000000 00000000 11110000 11110000` and `get_sub_group().get_local_linear_id() == 13`,
+ext_oneapi_rank_in_mask would return 5.
+
+
+```c++
+namespace sycl {
+template <int dimensions = 1>
+class nd_item {
+  public:
+...
+  // create a bitmask of the sub-group subset this work-item would be a part of
+  // if the sub-group was split into contiguous groups of size `partition_size`
+  sub_group_mask ext_oneapi_partition_sub_group(size_t partition_size) const;
+
+  // get the bitmask representing the work-items of the sub-group that are
+  // actively executing.
+  sub_group_mask ext_oneapi_active_sub_group_items() const;
+
+  //Returns an integer id representing the calling work-item's position within the provided sub_group_mask.
+  size_t ext_oneapi_rank_in_mask(sub_group_mask mask) const;
+...
+};
+}
+```
+
+possible implementation of `ext_oneapi_rank_in_mask`:
+```c++
+size_t ext_oneapi_rank_in_mask(sub_group_mask mask){
+  // taking 1 from a power of two will give all 1s below the bit of our work-item
+  // anding this with the mask will leave only 1s where there are work-items with lower ids
+  // the popcount of this is the linear id in that sub_group_mask
+  return popcount(mask & ((1 << get_sub_group().get_local_linear_id()) -1));
+}
+```
+
+## Changes to `joint_reduce` and `reduce_over_group`
+Addition of bitmasked version of methods for sub-groups.
+
+```c++
+namespace sycl::ext::oneapi {
+template <typename Ptr, class BinaryOperation>
+joint_reduce(sycl::sub_group, Ptr first, Ptr last, BinaryOperation binary_op, sycl::sub_group_mask);
+
+template <typename Ptr, typename T, class BinaryOperation>
+joint_reduce(sycl::sub_group, Ptr first, Ptr last, T init, BinaryOperation binary_op, sycl::sub_group_mask);
+
+template <typename T, class BinaryOperation>
+reduce_over_group(sycl::sub_group, T x, BinaryOperation binary_op, sycl::sub_group_mask);
+
+template <typename V, typename T, class BinaryOperation>
+reduce_over_group(sycl::sub_group, V x, T init, BinaryOperation binary_op, sycl::sub_group_mask);
+} // namespace sycl::ext::oneapi
+```
+
+## Group_barrier
+New overloads will be required for `sub_group` that takes a bitmask.
+
+```c++
+namespace sycl::ext::oneapi {
+group_barrier(sycl::sub_group, sycl::sub_group_mask mask, memory_scope fence_scope = sub_group::fence_scope);
```

Comment:
It might be a little surprising that `sycl::sub_group_mask` is the last non-defaulted argument here, as it might be harder for developers to remember where the argument should go.  A code might reasonably look like this:
```c++
sum = joint_reduce(group, first, last, std::plus<>(), mask);
/* do something with sum */
group_barrier(group, mask, memory_scope::device);
```

There might be less cognitive burden if the mask was always the second argument after the `sycl::sub_group`.  I haven't fully thought through if that could lead to ambiguity, though -- I suppose somebody might try to run an algorithm using a mask as an input value?

Reply:
I guess this would be another example of where a new group class or two could be useful.